src/bin/cargo-flamegraph.rs

```diff
@@ -174,91 +176,86 @@ fn build(opt: &Opt) {
         cmd.arg("--no-default-features");
     }
 
+    cmd.arg("--message-format=json");
+
     if opt.verbose {
         println!("build command: {:?}", cmd);
     }
 
-    let mut child = cmd
-        .spawn()
-        .expect("failed to spawn cargo build command");
-
-    let exit_status = child.wait().expect(
-        "failed to wait for cargo build child to finish",
-    );
+    let Output {
+        status,
+        stdout,
+        stderr,
+    } = cmd
+        .output()
+        .expect("failed to execute cargo build command");
+
+    let messages = Message::parse_stream(&*stdout);
+    let artifacts: Vec<_> = messages
+        .filter_map(|m| match m {
+            Ok(Message::CompilerArtifact(artifact)) => {
+                Some(artifact)
+            }
+            Ok(_) => None,
+            Err(e) => {
+                panic!("failed to parse cargo build output: {:?}", e);
+            }
+        })
+        .collect();
 
     if !opt.dev {
-        cmd.arg("--message-format=json");
+        check_debug_info(opt, &artifacts);
+    }
 
-        let output: Vec<u8> = cmd
-            .output()
-            .expect("failed to execute cargo build command")
-            .stdout;
+    if !status.success() {
+        eprintln!("cargo build failed: {:?}", stderr);
+        std::process::exit(1);
+    }
 
-        let messages =
-            cargo_metadata::Message::parse_stream(&*output);
+    artifacts
+}
 
-        let mut has_debuginfo = false;
+fn check_debug_info(opt: &Opt, artifacts: &[Artifact]) {
```

Comment:
Let's not create a single-use function here, and keep the checking of debug info inside `build()` as before. In general, this kind of refactoring should also go in a separate commit to make it easier to review.